openhands/server/session/manager.py

```diff
@@ -40,3 +80,97 @@ async def attach_to_conversation(self, sid: str) -> Conversation | None:
 
     async def detach_from_conversation(self, conversation: Conversation):
         await conversation.disconnect()
+
+    async def init_or_join_session(self, sid: str, connection_id: str, data: dict):
+        await self.sio.enter_room(connection_id, ROOM_KEY.format(sid=sid))
+        self.local_connection_id_to_session_id[connection_id] = sid
+        
+        # If we have a local session running, use that
+        session = self.local_sessions_by_sid.get(sid)
+        if session:
+            return session.agent_session.event_stream        
+
+        # If there is a remote session running, mark a connection to that
+        redis_client = self._get_redis_client()
+        if redis_client:
+            num_connections = await redis_client.rpush(_CONNECTION_KEY.format(sid=sid), connection_id)
+            # More than one remote connection implies session is already running remotely...
+            if num_connections != 1:
+                event_stream = EventStream(sid, self.file_store)
+                return event_stream
+
+        # Start a new local session
+        session = Session(
+            sid=sid, file_store=self.file_store, config=self.config, sio=self.sio
+        )
+        self.local_sessions_by_sid[sid] = session
+        await session.initialize_agent(data)
+        return session.agent_session.event_stream
+    
+    
+    async def send_to_event_stream(self, connection_id: str, data: dict):
+        # If there is a local session running, send to that
+        sid = self.local_connection_id_to_session_id[connection_id]
+        session = self.local_sessions_by_sid.get(sid)
+        if session:
+            await session.dispatch(data)
+            return
+        
+        # If there is a remote session running, send to that
+        redis_client = self._get_redis_client()
+        if redis_client:
+            await redis_client.publish("oh_event", json.dumps({
+                "sid": sid,
+                "data": data
+            }))
+            return
```

Comment:
I'm surprised we have to manually poke redis here and elsewhere in this file--shouldn't socketio handle this plumbing for us?